src/llmcompressor/modifiers/quantization/awq/base.py

```diff
@@ -0,0 +1,467 @@
+    balanced to offset the smoothing of smooth_layer
+    """
+
+    smooth_name: str
+    smooth_layer: Module
+    balance_layers: List[Module]
+
+
+class AWQModifier(Modifier):
+    """
+     Implements the AWQ algorithm from https://arxiv.org/pdf/2306.00978. This
+     modifier performs a smoothing of outliers in activations, making them
+     easier to quantize by reducing the dynamic range. The smoothing is offset by
+     applying the inverse operation to the next layer of weights, making the weights
+     slightly more difficult to quantize.
+
+     Because this modifier manipulates the weights of the model, it can only be used in
+     in one-shot and not during training. Activation ranges are determined by running a
+     small set of calibration data through the model.
+
+    example recipe:
+     ```yaml
+     AWQModifier:
+       bits: 4
+       mappings: [
+         [["re:.*q_proj", "re:.*k_proj", "re:.*v_proj"], "re:.*self_attn_layer_norm"],
+         [["re:.*fc1"], "re:.*final_layer_norm"]
+       ]
+       ignore: ["model.decoder.final_layer_norm"]
+     ```
+
+     :param mappings: list activation layers to smooth, and which layers to
+        scale the output such that activations are smoothed.
+        Each entry of the mapping list should be a list itself, in which the first
+        entry is a list of layers who share the same input activation (the one to be
+        to smoothed) and the second entry is the layer whose output is scaled to
+        achieve the smoothing.
+        If regex is used, it matches layers with the largest overlap in module name.
+     :param ignore: list of layers to ignore, even if they match a regex in mappings.
+        It should match the name of layers whose outputs are scaled to achieve
+        smoothing (the second entry of the mappings list).
+     :param num_calibration_steps: number of samples to use for calibration, or None to
+     use the whole dataset
+    :param calibration_function: optional function to use for the forward pass, or None
+    to use the default tensor_module_forward
+    """
+
+    mappings: List[Tuple] = DEFAULT_AWQ_MAPPINGS
+    ignore: Optional[List[str]] = None
+    num_calibration_steps: Optional[int] = None
+    calibration_function: Optional[Callable] = None
+    group_size: int = 128
+    max_chunk_memory: int = (1024 * 1024 * 1024,)
+    bits: int = 4
+    symmetric: bool = True
+
+    hooks_: Optional[List] = None
+    resolved_mappings_: Optional[List] = None
+    scales_: Optional[Dict] = None
+
+    def on_initialize_structure(self, state: State, **kwargs):
+        pass  # nothing needed for this modifier
+
+    def on_initialize(self, state: State, **kwargs) -> bool:
+        """
+        Initialize and run AWQ on the given state
+
+        :param state: state to run AWQ on
+        :return: True on a successful run, False otherwise
+        """
+        if self.end and self.end != -1:
```

Comment:
Currently user can pass self.end == 0
```suggestion
        if self.end is not None or self.end == -1:
```